src/hotspot/cpu/x86/gc/z/zBarrierSetAssembler_x86.cpp

```diff
@@ -78,15 +78,43 @@ class ZRuntimeCallSpill {
 
   void save() {
     MacroAssembler* masm = _masm;
-    __ push(rax);
-    __ push(rcx);
-    __ push(rdx);
-    __ push(rdi);
-    __ push(rsi);
-    __ push(r8);
-    __ push(r9);
-    __ push(r10);
-    __ push(r11);
+    if (VM_Version::supports_apx_f()) {
+      __ push(rax);
+      __ push(rcx);
+      // Save current stack pointer into rcx
+      __ movptr(rcx, rsp);
+      // Align stack pointer to 16 byte boundary. This is hard constraint
+      // for push2/pop2 with PPX hints.
+      __ andptr(rsp, -StackAlignmentInBytes);
+      // Push original stack pointer along with remaining registers on 16B aligned stack.
+      // Note: For PPX to work properly, a PPX-marked PUSH2 (respectively, POP2) should always
+      // be matched with a PPX-marked POP2 (PUSH2), not with two PPX-marked POPs (PUSHs).
+      __ pushp(rcx);
```

Comment:
This is saving old rsp on stack and restored using __ movptr(rsp, Address(rsp)) on the other end in restore().  So this should be __ push(rcx)  and not __ pushp(rcx) as there is no corresponding __ popp() instruction for this pushp.